upi/openstack/library/os_create_svc_subnet.py

```diff
@@ -0,0 +1,123 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+DOCUMENTATION = '''
+---
+module: os_create_svc_subnet
+short_description: Create a subnet
+description:
+    - Create a subnet for the services
+options:
+   subnet_name:
+     description:
+       - The name of the subnet to be create
+     required: true
+   network_name:
+     description:
+        - Network this subnet belongs to
+     required: true
+   cidr:
+     description:
+       - Subnet range in CIDR notation
+     required: true
+   openstack_svc_network:
+     description:
+        -  CIDR of network to allocate IPs for OpenStack
+           Octavia's Amphora VMs. Note that the value of this
+           option must be at least two times the size of the
+           cidr option, as octavia uses two IPs from this network
+           for each loadbalancer, one given by OpenShift and other
+           for VRRP connections.
+     required: true
+   ip_version:
+     description:
+       - IP version
+     required: true
+   enable_dhcp:
+     description:
+        - Enable DHCP
+     required: true
+'''
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.openstack import openstack_full_argument_spec, openstack_module_kwargs, openstack_cloud_from_module
+
+import ipaddress
+
+def main():
+    ''' Main module function '''
+    argument_spec = openstack_full_argument_spec(
+        subnet_name=dict(type='str', required=True),
+        network_name=dict(type='str', required=True),
+        cidr=dict(type='str', required=True),
+        openstack_svc_network=dict(type='str', required=True),
+        ip_version=dict(type='int', required=True),
+        enable_dhcp=dict(type=bool, required=True))
+
+    module_kwargs = openstack_module_kwargs()
+    module = AnsibleModule(argument_spec, **module_kwargs)
+
+    subnet_name = module.params['subnet_name']
+    network_name = module.params['network_name']
+    cidr = module.params['cidr']
+    openstack_cidr = module.params['openstack_svc_network']
+    ip_version = module.params['ip_version']
+    enable_dhcp = module.params['enable_dhcp']
+
+    svc_network = ipaddress.ip_network(cidr)
+    openstack_svc_cidr = ipaddress.ip_network(openstack_cidr)
+
+    # Parts of openstack_svc_cidr not overlapping with svc_network
+    allocation_ranges = []
+    # If the last IPs are different the first
+    # part of the openstack network is already allocated
+    if svc_network[-1] != openstack_svc_cidr[-1]:
+        allocation_ranges.extend([{
+            "start": str(svc_network[-1]+1),
+            "end": str(openstack_svc_cidr[-2])}])
+
+    # If the first IPs are different the last part
+    # of the openstack network is already allocated
+    if svc_network[0] != openstack_svc_cidr[0]:
```

Comment:
should this be elif?

Reply:
No, to account when cidr is not on the beginning nor in the end of the range. One example:
https://github.com/openshift/cluster-network-operator/blob/2dbafa8dcb09fbda20f5304e072fbbb25d20d970/pkg/util/ip/addr_test.go#L302